### ISSUE TITLE
Integrate CI integration testing into Github

### ISSUE THREAD

**Issue by bob2517** [Active-CSS/active-css#22]
All the unit testing currently happens offline and is integrated into the docs website admin area. I just click a button and it rattles off a bunch of tests. It's really cool - you can simulate real clicks, etc. in Active CSS. But as the admin area of the website isn't even on the internet, the unit testing isn't particularly incredibly useful on my local server. It needs to be part of the core dev suite so if anyone else other than me wants to hack the core, they can also do unit tests to make sure things aren't breaking.

The reason I didn't set it up a third-party test suite was that it was easier to just use Active CSS to trigger actions and verify with native JavaScript functions. But this method isn't necessarily going to translate to a commonly agreed workflow.

This will be done right after once the core is stable and I stop getting urgent implementation ideas, which is definitely getting closer. I dunno much about third-party testing suites or setting up continuous integration, so will need to investigate and make a decision on which way to go when the time arrives. The test framework chosen could ideally have performance statistics so that before and after measurements can be noted.

**Comment by bob2517** [comment]
Looking at using Karma. This looks like it ticks the box for automated cross-browser testing. Integrates with GitHub too.

**Comment by bob2517** [comment]
Will use Jasmine as the testing framework. WIll be setting up CI for the next release, but probably won't have all the tests ported over, if any at all. Need to get a release out and don't want to delay for too long. This is very needed though, so putting in the time to get this done now.

**Comment by bob2517** [comment]
Gonna use Jenkins for the CI.

**Comment by bob2517** [comment]
Active CSS repo is set up with Travis. Build passed, as to be expected, so it compiled without errors, but there aren't any tests yet. Next is to get Karma and Jasmine working with my offline grunt and some tests ported over.

**Comment by bob2517** [comment]
Karma, jasmine and phantomjs are set up with dummy tests. Next is to port the existing tests over.

**Comment by bob2517** [comment]
It's such a waste of time trying to set up old-ish Node packages that aren't fully documented. PhantomJS doesn't support ES6, and there's no mention of this in the Karma docs. Luckily there's a shim. Node dependencies are a nice idea in theory, it gets literally everyone involved, but you just waste so much time in setup in the long term as things go out of date. The best solution will always be a fully integrated system.

**Comment by bob2517** [comment]
Doh - looks like I might need Babel with the shim. Sigh.

**Comment by bob2517** [comment]
Babel has been put back into core building, as it's needed in order to translate the core for unit testing in Karma. Slightly concerned that it may not support shadow DOM when it comes to it. Will use Chrome or Firefox if need be.

**Comment by bob2517** [comment]
FirefoxHeadless being used now. Don't need Babel any more as that was a symptom of PhantomJS not supporting ES6.

**Comment by bob2517** [comment]
ChromeHeadless too.
No other browsers are particularly sensible to use right now. May look at an integration with browserstack at some point. For now though these are fine.

**Comment by bob2517** [comment]
As Active CSS is event-driven, all tests are going to be triggered by events. Helpers functions are set up.
Example of first successful test:

```
describe('Basic render', () => {
	beforeAll(function(){
		jasmine.addMatchers(DOMCustomMatchers);
	});

  	it('Draw a container div after a click', () => {
		doClick('body');
		let el = getObj('#content');
		expect(el).toBeDocumentNode();
	});

});
```

Where the test config is:
```
body:init {
	render: "<div id="content"></div>";
}
```

**Comment by bob2517** [comment]
No sooner that I write the last comment that I start debating whether or not just to use Active CSS to run the tests and just use Jasmine to start the process and check for green flags from Active CSS once all the tests have been run. I guess it doesn't matter as long as it's automated as part of the CI pipeline and it accurately displays errors. That is the whole point after all. It certainly would be quicker to write the tests from Active CSS, and it's going to be a lot easier to test shadow DOMs, all the custom events, etc.
Hmmm... gonna ponder this. Should speed things up quite a bit. Will be back on this at the weekend and run a few tests, so to speak.

**Comment by bob2517** [comment]
New for version 2.4.0 - ActiveCSSInitialized custom event will be fired after all initialization steps are completed, so that's the config loaded and all the init events are run. I need this for the tests and other people might find it useful so it's going in the core.

**Comment by bob2517** [comment]
Finally got the flow simple enough to get the tests ported over. Tests are going to be result-based. Why? Because Active CSS is not a JavaScript framework - it is a language. Because no-one ever calls the internals of Active CSS directly, I can change things around a lot in the core, changing parameters, etc, and having to change tests every time I do this will just break up my workflow and be pointless. All I am interested in is whether all the commands work and if the behaviours work as expected. That is all the developer is interested in too.

So all tests will be based on behaviour and commands, written in Active CSS, verified in native JS and marked as successful from there.

Jasmine is being used to make sure all the tests are marked as successful and will fail if any tests fail, displaying the failing tests in a list.

To create a test, two files are needed.

1) An Active CSS config file, rendering everything it needs to perform the test, plus the specific command or behaviour that is being tested. Every test need a wrapping div with a class of "coreTest". And every single test is triggered by the "content:draw" event. In Active CSS, you manually call the function you need to verify the test, so it can be named anything. The whole thing is done in Active CSS for simplicity. It doesn't go back and forth with Jasmine or anything. So all the tests run at once in Active CSS. When everything is finished, Jasmine takes over. Full instructions will be somewhere when released.

2) A Javascript verification file, that checks that the test is successful and if so, marking the container div with a class of successful.

Then you just need to run grunt to compile the core and run the tests.

Eg. This is the remove class test.

File: checkRemoveClass.acss
```
#content:draw {
    render-before-end: "<div id=\"checkRemoveClass\" data-desc=\"Remove-class command\" class=\"coreTest removeClassToRemove\"></div>";
}

#checkRemoveClass:test {
    remove-class: .removeClassToRemove;
    func: checkRemoveClass;
}
```

File: checkRemoveClass.js
```
function checkRemoveClass(o) {
    // Check that the element to remove the class from is definitely there.
    let el = _getObj('#checkRemoveClass');
    if (!el) {
        console.log('Unable to test checkRemoveClass as render isn\'t working.');
        return;
    }
    // Check if the class is no longer there.
    if (_hasClassObj(el, 'removeClassToRemove')) {
        // Test failed. No success marker gets added.
        return;
    }
    _addSuccessClass(el);
}
```

I think that's simple enough for anyone to get their head around.

**Comment by bob2517** [comment]
I'll probably add more helper functions as I go through the test compiling and will document these somewhere when 2.4.0 goes up.

**Comment by bob2517** [comment]
This is quite a meaty issue and will probably take a few weeks to complete, but once this is done it will speed development up a bit, so it should all balance out fine. It's one of those things that just has to get done and the longer I leave it the bigger the task to address.

**Comment by bob2517** [comment]
Example of successful grunt run with the two tests currently there:
(ASCII art courtesy of Joan Stark, https://www.asciiart.eu/computers/smileys)

```
Running "concat:devEdition" (concat) task

Running "concat:prodEdition" (concat) task

Running "concat:testingJS" (concat) task

Running "concat:testingConfig" (concat) task

Running "jshint:files" (jshint) task
>> 3 files lint free.

Running "terser:regular" (terser) task
>> 2 grunt.util.pluralize(createdFiles, 'file/files') created.

Running "karma:tests" (karma) task
06 10 2020 16:06:43.806:INFO [karma-server]: Karma v5.2.3 server started at http://localhost:9876/
06 10 2020 16:06:43.814:INFO [launcher]: Launching browsers ChromeHeadless with concurrency 1
06 10 2020 16:06:43.828:INFO [launcher]: Starting browser ChromeHeadless
06 10 2020 16:06:44.982:INFO [Chrome Headless 85.0.4183.121 (Linux i686)]: Connected on socket rL1EIcf2Qdns2L8aAAAA with id 61020505
LOG: '

The following tests were run:

Add-class command
Remove-class command

And all tests passed!

$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$'               `$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
$$$$$$$$$$$$$$$$$$$$$$$$$$$$'                   `$$$$$$$$$$$$$$$$$$$$$$$$$$$$
$$$'`$$$$$$$$$$$$$'`$$$$$$!                       !$$$$$$'`$$$$$$$$$$$$$'`$$$
$$$$  $$$$$$$$$$$  $$$$$$$                         $$$$$$$  $$$$$$$$$$$  $$$$
$$$$. `$' \' \$`  $$$$$$$!                         !$$$$$$$  '$/ `/ `$' .$$$$
$$$$$. !\  i  i .$$$$$$$$                           $$$$$$$$. i  i  /! .$$$$$
$$$$$$   `--`--.$$$$$$$$$                           $$$$$$$$$.--'--'   $$$$$$
$$$$$$L        `$$$$$^^$$                           $$^^$$$$$'        J$$$$$$
$$$$$$$.   .'   ""~   $$$    $.                 .$  $$$   ~""   `.   .$$$$$$$
$$$$$$$$.  ;      .e$$$$$!    $$.             .$$  !$$$$$e,      ;  .$$$$$$$$
$$$$$$$$$   `.$$$$$$$$$$$$     $$$.         .$$$   $$$$$$$$$$$$.'   $$$$$$$$$
$$$$$$$$    .$$$$$$$$$$$$$!     $$`$$$$$$$$'$$    !$$$$$$$$$$$$$.    $$$$$$$$
$JT&yd$     $$$$$$$$$$$$$$$$.    $    $$    $   .$$$$$$$$$$$$$$$$     $by&TL$
                                 $    $$    $
                                 $.   $$   .$
                                 `$        $'
                                  `$$$$$$$$'

'
Chrome Headless 85.0.4183.121 (Linux i686): Executed 1 of 1 SUCCESS (7.05 secs / 7.006 secs)
TOTAL: 1 SUCCESS

Done.
```

**Comment by bob2517** [comment]
Status report: Almost half-way through the command tests. Then it'll be the other integration tests, like shadow DOM and stuff like that. Nothing found yet to fix. Just a few commands that could do with some functionality upgrading which won't happen until after the next release.

**Comment by bob2517** [comment]
While working on removing the data-activeid attributes and reactive placeholders issue, it has become apparent that there needs to be some tests for memory leakage.

So this is being added to the list...

**Comment by bob2517** [comment]
Ok - back on this now. The plan now is to get all the commands and conditional tests written as a milestone. Then get any other urgent outstanding issues wrapped up for release, wrap up the docs, and then get the new release out. I've got enough tests set up offline that are not part of the CI suite to cover everything else for the time being. Then once the new release is out I'll get the remainder of the CI tests written.

**Comment by bob2517** [comment]
This is definitely being actively worked on - quite a number of new issues have come up and been handled as a result of getting the test suites done. Hence it doesn't look like much has happened but it has. Hopefully won't be too long before these test suites are done now. No other upgrades other than fixes are planned for 2.4.0.

**Comment by bob2517** [comment]
Just one more set of test suites to do and then that will be it for 2.4.0. Should get those wrapped by the end of Saturday if not sooner.

**Comment by bob2517** [comment]
All the basic integration tests are now on the branch that were planned for 2.4.0 release. Will be adding more tests over time. For now though, here endeth the test suite initial evolution.
